src/notification/index.js

```diff
@@ -71,23 +71,33 @@ export const getAccountFromNotificationData = (
     return null;
   }
 
-  if (urlMatches.length > 1) {
-    // The user has several accounts in the notification's realm.  We should
-    // be able to tell the right one using the notification's `user_id`...
-    // except we don't store user IDs in `accounts`, only emails.  Until we
-    // fix that, just ignore the information.
-    logging.warn('notification realm_uri ambiguous; multiple matches found', {
-      realm_uri,
-      parsed_url: realmUrl,
-      match_count: urlMatches.length,
-      unique_identities_count: new Set(urlMatches.map(matchIndex => identities[matchIndex].email))
-        .size,
-    });
-    // TODO get user_id into accounts data, and use that
-    return null;
+  // TODO(server-2.1): Remove this, because user_id will always be present
+  if (user_id === undefined) {
+    if (urlMatches.length > 1) {
+      logging.warn(
+        'notification realm_uri ambiguous; multiple matches found; user_id missing (old server)',
+        {
+          realm_uri,
+          parsed_url: realmUrl,
+          match_count: urlMatches.length,
+          unique_identities_count: new Set(urlMatches.map(matchIndex => accounts[matchIndex].email))
+            .size,
+        },
+      );
+      return null;
+    } else {
+      return urlMatches[0];
+    }
   }
 
-  return urlMatches[0];
+  // The user has several accounts in the notification's realm. Pick one
+  // based on the notification's `user_id`.
+  const userMatch = urlMatches.find(urlMatch => accounts[urlMatch].userId === user_id);
+  if (userMatch == null) {
+    logging.warn('notification account not found with realm_uri and user_id');
```

Comment:
Although at this point we have a `user_id` from the notification, it may be the case that the account it's meant for doesn't have `userId` non-null in the accounts state. That's because:
```js
  /**
   * The user's numeric ID.
   *
   * We learn the user ID each time we complete an initial fetch.
   *
   * This is `null` briefly when we've logged in but not yet completed our
   * first initial fetch on the account.  It's also `null` when representing
   * an account which was last used on a version of the app which didn't
   * record this information.  It's never `null` for an account for which we
   * have server data.
   */
```

Probably if there's just one URL match, and its `userId` is null, then we should go for it.

Reply:
> Probably if there's just one URL match, and its `userId` is null, then we should go for it.

My current revision uses a slightly different algorithm. If there are two URL matches, and one of them has a `userId` but it's the wrong one, but the other one is `null`, we still want to go for the `null` one.